src/H5Dio.c

```diff
@@ -429,7 +449,7 @@ H5D__read(size_t count, H5D_dset_io_info_t *dset_info)
 
 done:
     /* Shut down the I/O op information */
-    for (i = 0; i < io_op_init; i++)
+    for (i = 0; i < count; i++)
```

Comment:
Here, we need to step through the full `count` number of dataset entries and call their `io_term` callbacks, instead of just `io_op_init` number of them. Otherwise, if the first `io_op_init` entries were skipped datasets, the latter entries won't get terminated and certain things are left around in memory, such as the cached dataset `sel_pieces` array. Unintended reuse of that cached field when doing an `H5Dread_multi` followed by `H5Dwrite_multi` gave me errors like:

```
HDF5-DIAG: Error detected in HDF5 (1.15.0) MPI-process 0:
  #000: /home/jhenderson/git/hdf5/src/H5D.c line 1446 in H5Dwrite_multi(): can't synchronously write data
    major: Dataset
    minor: Write failed
  #001: /home/jhenderson/git/hdf5/src/H5D.c line 1317 in H5D__write_api_common(): can't write data
    major: Dataset
    minor: Write failed
  #002: /home/jhenderson/git/hdf5/src/H5VLcallback.c line 2282 in H5VL_dataset_write_direct(): dataset write failed
    major: Virtual Object Layer
    minor: Write failed
  #003: /home/jhenderson/git/hdf5/src/H5VLcallback.c line 2237 in H5VL__dataset_write(): dataset write failed
    major: Virtual Object Layer
    minor: Write failed
  #004: /home/jhenderson/git/hdf5/src/H5VLnative_dataset.c line 420 in H5VL__native_dataset_write(): can't write data
    major: Dataset
    minor: Write failed
  #005: /home/jhenderson/git/hdf5/src/H5Dio.c line 732 in H5D__write(): can't initialize I/O info
    major: Dataset
    minor: Unable to initialize object
  #006: /home/jhenderson/git/hdf5/src/H5Dchunk.c line 1091 in H5D__chunk_io_init(): unable to create file and memory chunk selections
    major: Dataset
    minor: Unable to initialize object
  #007: /home/jhenderson/git/hdf5/src/H5Dchunk.c line 1241 in H5D__chunk_io_init_selections(): unable to create file chunk selections
    major: Dataset
    minor: Unable to initialize object
  #008: /home/jhenderson/git/hdf5/src/H5Dchunk.c line 1944 in H5D__create_piece_file_map_hyper(): can't insert piece into skip list
    major: Dataspace
    minor: Unable to insert object
  #009: /home/jhenderson/git/hdf5/src/H5SL.c line 1036 in H5SL_insert(): can't create new skip list node
    major: Skip Lists
    minor: Unable to insert object
  #010: /home/jhenderson/git/hdf5/src/H5SL.c line 709 in H5SL__insert_common(): can't insert duplicate key
    major: Skip Lists
    minor: Unable to insert object
```